### ISSUE TITLE
Apparent miscompilations

### ISSUE THREAD

**Issue by tari** [Jonimoose/libfxcg#42]
tswilliamson has reported issues with the new builds, where the binaries crash when they don't if built with the old SDK package.

Provided a large archive of the workspace state with the new SDK for building nesizm, which exhibits issues. Too large to attach here sadly. We can compare against what I think is the reference build via https://github.com/tswilliamson/nesizm and https://github.com/tswilliamson/PrizmSDK

Reports the binary dies on startup with a weird PC value, so hopefully this isn't too hard of an issue to debug.

**Comment by tari** [comment]
The nesizm g3a doesn't link when I try to build it, but here are the object files, in both debug and release mode for the 0.5 release (-0.5/) and tswilliamson's fork (-tsw/) of the SDK.

[builds.zip](https://github.com/Jonimoose/libfxcg/files/4692190/builds.zip)

**Comment by tari** [comment]
Crash symptoms:

> It wasn't ever able to get into main()
>
> I did some digging for you. It is crashing at PC=0x0810946C TARGET=0x0031C36B, which from the output.map file ([output.map.txt](https://github.com/Jonimoose/libfxcg/files/4698480/output.map.txt)) is initial ctors, possibly stdout:

```
.data._impl_stdout
0x0810945c 0x10 C:\Users\Thomas\Documents\Prizm\PrizmSDK-Tari\projects\nesizm\proj\/../../..//lib\libc.a(stdio.o)
0x0810945c __impl_stdout
0x0810946c _edata = .

.ctors 0x0810946c 0xc load address 0x003257d3
.ctors 0x0810946c 0x4 frontend.o
.ctors 0x08109470 0x4 nes_apu.o
.ctors 0x08109474 0x4 nes_cart.o
```

> 0x31C36B is a bit more cryptic:

```
.data 0x08100004 0x9468 load address 0x0031c36b
0x0031c36b _datald = LOADADDR (.data)
0x08100004 _sdata = .
```

> My best two guesses: either some data area is wrong, OR this is a 4-byte alignment issue that the compiler is not protecting against with its symbol placement.

**Comment by tari** [comment]
So `.ctors` is a compiler-generated section for functions that should be called in startup; it's a `-1`-terminated list of function pointers. Where is that section coming from? Our linker script doesn't include it.

Linker map says the first entry comes from frontend.o, so what's that a ref to?

```
$  objdump projects/nesizm/DeviceDebug/frontend.o -s -r -j .ctors
projects/nesizm/DeviceDebug/frontend.o:     file format elf32-big

RELOCATION RECORDS FOR [.ctors]:
OFFSET   TYPE              VALUE
00000000 UNKNOWN           .text.startup


Contents of section .ctors:
 0000 00000000
```

Okay, it contains a ref to `.text.startup` in the same object file.

```
$ objdump projects/nesizm/DeviceDebug/frontend.o -s -r -j .text.startup
projects/nesizm/DeviceDebug/frontend.o:     file format elf32-big

RELOCATION RECORDS FOR [.text.startup]:
OFFSET   TYPE              VALUE
00000010 UNKNOWN           _nesFrontend


Contents of section .text.startup:
 0000 d103e200 21221121 11221123 000b1124  ....!".!.".#...$
 0010 00000000
```

That makes sense; `frontend.cpp` contains a global `nes_frontend nesFrontend` which has a constructor, so the compiler puts a ref to its constructor in there to be run at startup.

Open questions:
 * How is the ctors section ending up in the binary? (Or is it?)
 * Does the problem go away if static constructors aren't used?

**Comment by tari** [comment]
It's also suspicious that _datald is not 32-bit aligned however, since crt0 does word-aligned initialization of the data section.

I think I need to disassemble the g3a and see what it's actually running on startup.

**Comment by tari** [comment]
Loading up a g3a, I immediately see that the data section VMA is 0x31C95F which is not aligned. Maybe this is as easy as aligning the RAM sections (both VMA and LMA).

```diff
diff --git a/toolchain/prizm.x b/toolchain/prizm.x
index 9b64317..0d15d6d 100755
--- a/toolchain/prizm.x
+++ b/toolchain/prizm.x
@@ -36,7 +36,7 @@ SECTIONS
         } > rom

         /* RW initialized data, VMA in RAM but LMA in ROM */
-        .data : {
+        .data : ALIGN(4) {
                 _datald = LOADADDR(.data) ;
                 _sdata = . ;
                 *(.data)
@@ -45,7 +45,7 @@ SECTIONS
         } >ram AT>rom

         /* Uninitialized data (fill with 0), in RAM */
-        .bss : {
+        .bss : ALIGN(4) {
                 _bbss = . ;
                 *(.bss)
                 *(.bss*)
```

**Comment by tari** [comment]
Fixed by f374fcaaae1b77bd0b259d06cd3b00522cbccdff.
